### PR TITLE
[apache_tomcat.access] Restructure Grok pattern and ingest pipeline to improve performance

### PR DESCRIPTION
**Pull request by SimonAtElastic**
The customer is using this Integration on an ESS and had ~10-14ms per doc average processing Time.
image
After taking a look at the pipeline definition I recognised the "inefficient" way how the grok processing is done.
First test of restructuring the grok process (without changing what it is doing, just how it is done) I had an average of 0-5-0.-7ms, so a huge performance improvement with just a little restructuring.
image

### CODE REVIEW

File: packages/apache_tomcat/manifest.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 format_version: "3.0.0"
 name: apache_tomcat
 title: Apache Tomcat
-version: "1.0.0"
+version: "1.0.1"
```
Comment (ishleenk17) [inline packages/apache_tomcat/manifest.yml:4]: Since its an enhancement, lets have it as 1.1.0

File: packages/apache_tomcat/changelog.yml [hunk 2]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "1.0.1"
+  changes:
+    - description: Improve apacht_tomcat.access pipeline performance
```
Comment (ishleenk17) [inline packages/apache_tomcat/changelog.yml:4]:
```suggestion
    - description: Improve apache_tomcat.access pipeline performance
```

**Review by ishleenk17 — COMMENTED** [review]
Some nits. Otherwise the change looks good. Thanks

File: packages/apache_tomcat/data_stream/access/elasticsearch/ingest_pipeline/default.yml [hunk 3]
```diff
@@ -25,15 +25,27 @@ processors:
...
+        - '^(%{IP:source.ip}|%{DATA:source.user.name}) %{DATA:apache_tomcat.access.http.ident} %{DATA:apache_tomcat.access.http.useragent} \[%{DATA:_tmp.timestamp}\] \"%{DATA:http.request.method} %{DATA:url.original} HTTP\/%{DATA:http.version}\" %{NUMBER:http.response.status_code} %{POSINT:destination.bytes}( %{GREEDYDATA:_tmp_grok})?$'
+      on_failure:
+        - append:
+            field: error.message
+            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag fail-{{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
+  - grok:
+      field: _tmp_grok
+      tag: 'grok_parse_log2'
+      patterns:
+        - '(?:%{IP:apache_tomcat.access.ip.local}%{SPACE})?(?:%{CONN_STATUS:apache_tomcat.access.connection_status}%{SPACE})?(?:%{NUMBER:apache_tomcat.access.response_time}%{SPACE})?\"%{DATA:http.request.referrer}\" \"%{DATA:user_agent.original}\" X-Forwarded-For=\"%{DATA:_tmp.header_forwarder}(\")*$'
```
Comment (shmsr) [inline packages/apache_tomcat/data_stream/access/elasticsearch/ingest_pipeline/default.yml:37]: Shouldn't `X-Forwarded-For ...` be in an optional group? For this one:
```
'^(%{IP:source.ip}|%{DATA:source.user.name}) %{DATA:apache_tomcat.access.http.ident} %{DATA:apache_tomcat.access.http.useragent} \[%{DATA:_tmp.timestamp}\] \"%{DATA:http.request.method} %{DATA:url.original} HTTP\/%{DATA:http.version}\" %{NUMBER:http.response.status_code} %{DATA:destination.bytes} \"%{DATA:http.request.referrer}\" \"%{DATA:user_agent.original}(\")*$'
```

File: packages/apache_tomcat/data_stream/access/elasticsearch/ingest_pipeline/default.yml [hunk 4]
```diff
@@ -25,15 +25,27 @@ processors:
...
-        - '^(%{IP:source.ip}|%{DATA:source.user.name}) %{DATA:apache_tomcat.access.http.ident} %{DATA:apache_tomcat.access.http.useragent} \[%{DATA:_tmp.timestamp}\] \"%{DATA:http.request.method} %{DATA:url.original} HTTP\/%{DATA:http.version}\" %{NUMBER:http.response.status_code} %{DATA:destination.bytes} \"%{DATA:http.request.referrer}\" \"%{DATA:user_agent.original}(\")*$'
-        - '^(%{IP:source.ip}|%{DATA:source.user.name}) %{DATA:apache_tomcat.access.http.ident} %{DATA:apache_tomcat.access.http.useragent} \[%{DATA:_tmp.timestamp}\] \"%{DATA:http.request.method} %{DATA:url.original} HTTP\/%{DATA:http.version}\" %{NUMBER:http.response.status_code} %{DATA:destination.bytes}$'
+        - '^(%{IP:source.ip}|%{DATA:source.user.name}) %{DATA:apache_tomcat.access.http.ident} %{DATA:apache_tomcat.access.http.useragent} \[%{DATA:_tmp.timestamp}\] \"%{DATA:http.request.method} %{DATA:url.original} HTTP\/%{DATA:http.version}\" %{NUMBER:http.response.status_code} %{POSINT:destination.bytes}( %{GREEDYDATA:_tmp_grok})?$'
```
Comment (shmsr) [inline packages/apache_tomcat/data_stream/access/elasticsearch/ingest_pipeline/default.yml:28]:
```suggestion
        - '^(%{IP:source.ip}|%{DATA:source.user.name}) %{DATA:apache_tomcat.access.http.ident} %{DATA:apache_tomcat.access.http.useragent} \[%{HTTPDATE:_tmp.timestamp}\] \"%{WORD:http.request.method} %{URI:url.original} HTTP\/%{DATA:http.version}\" %{NUMBER:http.response.status_code} %{POSINT:destination.bytes}( %{GREEDYDATA:_tmp_grok})?$'
```

I didn't test it but these changes could further improve perf? 

Ref: https://github.com/hpcugent/logstash-patterns/blob/master/files/grok-patterns

**Review by shmsr — CHANGES_REQUESTED** [review]

File: packages/apache_tomcat/data_stream/access/elasticsearch/ingest_pipeline/default.yml [hunk 5]
```diff
@@ -25,15 +25,27 @@ processors:
...
+      on_failure:
+        - append:
+            field: error.message
+            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag fail-{{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
+  - grok:
+      field: _tmp_grok
```
Comment (shmsr) [inline packages/apache_tomcat/data_stream/access/elasticsearch/ingest_pipeline/default.yml:34]: What if we rename this field to `_tmp.grok`? Then the `remove` processor for `_tmp` will delete it at once and 1 less call to remove will be made?

```
- remove:
      field:
        - _tmp
      ignore_missing: true
```

Already present.

File: packages/apache_tomcat/data_stream/access/elasticsearch/ingest_pipeline/default.yml (hunk 3 above)
Comment (SimonAtElastic) [reply packages/apache_tomcat/data_stream/access/elasticsearch/ingest_pipeline/default.yml:37]: not sure about that. As i don't know the logs in detail in enough to confidentaly answer this and the Integrations works so far, i would tend to keep it as is it so far.

**Review by SimonAtElastic — COMMENTED** [review]
fixed/comment all suggestions

**Review by shmsr — APPROVED** [review]

**Review by harnish-elastic — APPROVED** [review]
LGTM!

**Pull request merged**